### PR TITLE
Threads and isolate in flutter

### LINKED ISSUE #4136: Thread and Isolate in Flutter
#### Pre-submission advice
By following all our pre-submission advice and reviewing our Resources folder, you will maximise your chances of your topic being approved.

We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner.

Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles that will help you propose and write a successful article.

**Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

#### Thread and Isolate with Flutter
Your title should be descriptive of the article/tutorial. Be Specific. 

Use keyword research to improve your article's SEO.

#### Proposed article introduction
An isolate is a thread that has an event loop that continuously processes events in its own memory space. In effect when the main dart function executes, it executes in an isolate that scans for available events and hangs when none are available, an opportunity for the system to decide to run the garbage collector for example.
Dart can execute multiple threads by spawning additional isolates. Creating more isolates can be useful for heavy computation processing but most Flutter apps do not need spawning additional isolates unless it becomes an issue and impacts the frame rate.
For example, if the article is based on machine learning, use the following: [Machine learning] Introduction to Machine Learning.
If the article is based on developing an Android application, use the following: [Android] Developing Apps using Android.

ONLY viable topics to pick from:
- API
- Artificial Intelligence
- Blockchain
- Containers
- Edge Computing
- Languages
- Machine Learning
- Networking
- Node.js
- Security

#### Key takeaways 
- Understand how to use isolate with flutter
- Understand event loops in asynchronous way
- code the implementation of the thread and isolate

#### Article quality
How is your article unique? Tell us what makes your approach different from similar articles that have published on the same topic?
Is yours more in-depth? Does it cover additional topics? Do you provide handy tips or anecdotal advice?

#### References
-https://flutter.dev/docs/development/packages-and-plugins/background-processes
- https://medium.com/flutter-community/thread-and-isolate-with-flutter-30b9631137f3

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by waigwakanoi**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #4136

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by hectorkambow** [comment]
Good afternoon and thank you for submitting your article for publishing.
Your article has been entered into our queue and should be reviewed (and published pending any changes) as soon as a peer reviewer is finished with the ones in the queue before it. 🚀
Thank you.

**Review by BrianaNzivu — CHANGES_REQUESTED** [review]
Hey @waigwakanoi  👋 .

First of all, welcome to Section - EngEd. :tada: 
We are happy to have you here.

On behalf of the team, I would love to apologize for the long wait in the queue. I promise to make this review as fast and as swift as possible, with your help of course 😄 .

Good job on the article. It is a very interesting read. Barely any grammar changes to be made.
This is very impressive considering you are a first-time contributor. Keep up the good work 👍 .

Just a few changes I need you to make as we continue with the preliminary revisions:
- Please structure your article according to this [sample](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md). I need you to clearly highlight and explain your steps. Where you can add more detail please do so.
- Scale down your avatar and hero image to 600 x 400 px as required. You can use [Befunky](https://www.befunky.com/) to do this.

In case of any question or help needed please do not hesitate to ping me here :smile:.
Also, notify me when you are done.

Regards.

**Comment by waigwakanoi** [comment]
@BrianaNzivu kindly reviews changes

**Comment by Kayere** [comment]
Plagiarism check done ✅

File: content/articles/threads-and-isolates-in-flutter/index.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
+### Threads and Isolates In Flutter Apps 
+### Introduction
+Threads and isolates are two of the essential abstractions you will need to know when developing apps in Flutter. They allow developers to utilize multicore processors, improve user experience by reducing UI stuttering, and reduce battery consumption on mobile devices. 
+### Table of contents
+- Introduction
```
Comment (BrianaNzivu) [inline content/articles/threads-and-isolates-in-flutter/index.md]: The table of contents has not been linked well. It is done this way:
[Introduction](#introduction)

Do the same for the rest.

File: content/articles/threads-and-isolates-in-flutter/index.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
+### Threads and Isolates In Flutter Apps 
+### Introduction
+Threads and isolates are two of the essential abstractions you will need to know when developing apps in Flutter. They allow developers to utilize multicore processors, improve user experience by reducing UI stuttering, and reduce battery consumption on mobile devices.
```
Comment (BrianaNzivu) [inline content/articles/threads-and-isolates-in-flutter/index.md]: Maybe you can define what threads and isolates are.

File: content/articles/threads-and-isolates-in-flutter/index.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+You can use the Isolate API to spawn new isolates; you can pass messages between an isolate and its parent thread(s). However, you are not directly accessing shared mutable states from within an isolate.
+### Application Development 
+To demonstrate their usage, we will build a simple counter-demo: An application that starts with an initial value at 0 and allows the user to increment or decrement that value.
+
+The application has a button that will start a new counter. When the corresponding Isolate is created, we store its
+instance in a variable and pass it through various message handlers to update the current counter value. The instance of the isolated thread is stored in a class member variable called _isolate.
+
+### Step 1 - Creating States
+- In the following code snippets, a comment starting with // [START isolate] means that an isolated thread is started. A comment beginning with // [END isolate] means the corresponding isolation is terminated.
```
Comment (BrianaNzivu) [inline content/articles/threads-and-isolates-in-flutter/index.md]: You can explain this in a better way. Talk about in this step we will be creating the states. Talk about the methods being used and then complete with the statement you have written.

File: content/articles/threads-and-isolates-in-flutter/index.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  currentCounter : _counter + 1 , message : msg , isStarting : start ) ) ; 
+  // [END isolate]
+   } /// Sends a message to  update the counter with a value of newValue and /// with isStarting set to true if start parameter is non-null. void
+  incrementCounter ( MyMessage msg , int newValue , bool start ) { 
+    // [START isolate] 
+  _isolate . send ( new
+  IncrementCounter ( currentCounter : _counter + newValue , isStarting : start ) ) ; 
+  // [END isolate] }
+  ```
+### Step 4 - Serialization
```
Comment (BrianaNzivu) [inline content/articles/threads-and-isolates-in-flutter/index.md:129]: of what?

File: content/articles/threads-and-isolates-in-flutter/index.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+  } 
+  /// Adds a function that can be called from the main thread. 
+  void incrementCounter ( IncrementCounterProviderFunction callback ) { 
+    // [START isolate] 
+    _isolate . send ( new
+  IncrementCounterProviderFunction ( onMessage : callback . onMessage , messageType : 0 , dataType : 1 ) ) ; 
+  // [END  isolate] 
+  }
+  ```
+### Step 5 - Provide Function
```
Comment (BrianaNzivu) [inline content/articles/threads-and-isolates-in-flutter/index.md:172]: To what?

File: content/articles/threads-and-isolates-in-flutter/index.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+  We also provide a special type of function called a ProviderFunction.A ProviderFunction can be called from the main  thread, and it returns a non - Serializable value to the main thread. 
+  
+  This is useful for cases where you want to return an object or data which is not serialized by default (for example, a std::shared_ptr).
+
+  Here is an example:
+  
+  ```dart
+  /// Sends a message to update the counter with a value of newValue and /// with isStarting set to true if start
```
Comment (BrianaNzivu) [inline content/articles/threads-and-isolates-in-flutter/index.md:180]: Have this code properly indented

**Review by BrianaNzivu — CHANGES_REQUESTED** [review]
Hey @waigwakanoi 👋 
Just a few more corrections to be made.

Regards.

**Comment by BrianaNzivu** [comment]
Tech review done 👍

**Review by BrianaNzivu — APPROVED** [review]
Hey @waigwakanoi 👋 .

Very good job on the article. It is a very interesting read and has minimal mistakes. This is very impressive considering you are a first-time contributor 👏 .

Just a few things I'd like to point out:
- I suggest that you check out our resources page which has links to various resourceful tools for writing just to minimize the changes being made by the reviewer.
- Ensure you explain your steps clearly and also indent your code properly.
- Lastly please go through ALL the corrections I have made so that you can not repeat the same mistakes.

That's it! :smile: 

A final review will be done now by a content moderator.
Otherwise, congratulations on publishing your first article🍾.

In case of any question or clarification (currently or in the future), do not hesitate to reach out via [Slack](https://app.slack.com/client/T5X14SH7G/D01A6MT673R) or my personal email (briananzivu@gmail.com). 
I am **ALWAYS** here to help.

It has been a pleasure reading and reviewing this article.

Looking forward to seeing you contribute more to EngEd. 🚀

Once again good job! 👍 

Kind Regards.

**Comment by lalith1403** [comment]
Amazing job, @waigwakanoi and @BrianaNzivu. Thanks for the wonderful article. Keep up the fantastic work :)

**Comment by hectorkambow** [comment]
@waigwakanoi Please DM me via Slack asap. 

This article will not received a payout this period - we have no PayPal on file - and the cut off time for this period is up.

**Comment by hectorkambow** [comment]
Also you may want to consider changing your avatar image (to a more professional looking one) as these articles are intended to help student jump start their developer careers and may be seen by potential employers.

**Comment by waigwakanoi** [comment]
okay sir

**Pull request merged**